software/py/generate_stats.py

```diff
@@ -0,0 +1,227 @@
+    for i in range (0, self.num_tile_groups):
+      self.execution_stats_file.write("{:10}{:5}{:25}{}\n".format("Tile group", i, ":", self.timing_end_list[i] - self.timing_start_list[i]))
+      self.total_execution_time += (self.timing_end_list[i] - self.timing_start_list[i])
+    self.execution_stats_file.write("{:40}{}\n".format("Total (cycles):", self.total_execution_time))
+    self.execution_stats_file.write("=======================================================\n\n")
+
+
+  # Print instruction stats for all tiles and total
+  def print_stats_instructions(self):
+    self.execution_stats_file.write("Instruction Stats:\n" + \
+                                    "=======================================================\n")
+    for instr, cnt in self.stats_instr_dict.items():
+       if instr != 'instr':
+         self.total_instr_cnt += cnt
+     
+    for instr, cnt in self.stats_instr_dict.items():
+       self.execution_stats_file.write("{:35}%{:0>5.2f}{:10}\n".format(instr, (100 * cnt / self.total_instr_cnt), cnt))
+
+    self.execution_stats_file.write("{:41}{:10}\n".format("Total", self.total_instr_cnt))
+    self.execution_stats_file.write("=======================================================\n\n")
+    return
+
+
+  # Print stall stats for all tiles and total
+  def print_stats_stalls(self):
+    self.execution_stats_file.write("Stalls Stats:\n" + \
+                                    "=======================================================\n")
+    for stall, cnt in self.stats_stall_dict.items():
+         self.total_stall_cnt += cnt
+     
+    for stall, cnt in self.stats_stall_dict.items():
+       self.execution_stats_file.write("{:35}%{:0>5.2f}{:10}\n".format(stall, (100 * cnt / self.total_stall_cnt), cnt))
+
+    self.execution_stats_file.write("{:41}{:10}\n".format("Total", self.total_stall_cnt))
+    self.execution_stats_file.write("=======================================================\n\n")
+    return
+
+
+  # Print miss stats for all tiles and total
+  def print_stats_miss(self):
+    self.execution_stats_file.write("Miss Stats:\n" + \
+                                    "=======================================================\n")
+    for miss, cnt in self.stats_miss_dict.items():
+         self.total_miss_cnt += cnt
+     
+    for miss, cnt in self.stats_miss_dict.items():
+       self.execution_stats_file.write("{:35}%{:0>5.2f}{:10}\n".format(miss, (100 * cnt / self.total_miss_cnt), cnt))
+
+    self.execution_stats_file.write("{:41}{:10}\n".format("Total", self.total_miss_cnt))
+    self.execution_stats_file.write("=======================================================\n\n")
+    return
+
+
+  def print_stats_all(self):
+    self.print_stats_timing()
+    self.print_stats_instructions()
+    self.print_stats_stalls()
+    self.print_stats_miss()
+
+
+
+
+  # default stats generator
+  def generate_stats(self, input_file):
```

Comment:
I think there is a csv module for python.
https://docs.python.org/3.6/library/csv.html